components/places/src/bookmark_sync/incoming.rs

```diff
@@ -521,83 +471,122 @@ impl<'a> IncomingApplicator<'a> {
     }
 }
 
-fn unpack_id(key: &str, data: &JsonValue) -> Result<BookmarkRecordId> {
-    if let Some(id) = data[key].as_str() {
-        Ok(BookmarkRecordId::from_payload_id(id.into()))
-    } else {
-        Err(Error::InvalidPlaceInfo(InvalidPlaceInfo::InvalidGuid))
+/// Go through the raw JSON value and try to fixup invalid data -- this mostly means fields with
+/// invalid types.
+///
+/// This is extra important since bookmarks form a tree.  If a parent node is invalid, then we will
+/// have issues tring to merge its children.
+fn fixup_bookmark_json(json: &mut JsonValue) -> SyncedBookmarkValidity {
+    let mut validity = SyncedBookmarkValidity::Valid;
+    // the json value should always be on object, if not don't try to do any fixups.  The result will
+    // be that into_content_with_fixup() returns an IncomingContent with IncomingKind::Malformed.
+    if let JsonValue::Object(ref mut obj) = json {
+        obj.entry("parentid")
+            .and_modify(|v| fixup_optional_str(v, &mut validity));
```

Comment:
Before there was a an `unpack_optional_id` that worked like `unpack_optional_str`, but didn't modify the validity.  I don't see why we wouldn't want to re-upload when the `parentid` field is wrong, so I made them use the same codepath.